### PR TITLE
Clean MetaTileEntity_Scanner, a couple niceties with achievements and…

### PR DESCRIPTION
**Pull request by Elisis**
… main file

### CODE REVIEW

**Review by bartimaeusnek — CHANGES_REQUESTED** [review]
some packages of org.apache.commons can crash on servers, you changed visibility of some public variables and added final modifiers on public variables and removed some fields. This is a problem with the addons / general stability.

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java [hunk 1]
```diff
@@ -191,124 +239,139 @@ else if (aStack.getItemDamage() == 2 && aStack.toString().equals(GT_ModHandler.g
...
-                        String s = tRecipe.mOutput.getDisplayName();
-                        if (FMLCommonHandler.instance().getEffectiveSide().isServer()) {
-                            s = GT_Assemblyline_Server.lServerNames.get(tRecipe.mOutput.getDisplayName());
-                            if (s == null)
-                                s = tRecipe.mOutput.getDisplayName();
-                        }
+                        String outputDisplayName = tRecipe.mOutput.getDisplayName();
+                        if (FMLCommonHandler.instance().getEffectiveSide().isServer()) 
+                            outputDisplayName = ObjectUtils.defaultIfNull(
+                                    GT_Assemblyline_Server.lServerNames.get(tRecipe.mOutput.getDisplayName()),
```
Comment (Skiftcha) [inline src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java]: getOrDefault(outputDisplayName, outputDisplayName)

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java [hunk 2]
```diff
@@ -191,124 +239,139 @@ else if (aStack.getItemDamage() == 2 && aStack.toString().equals(GT_ModHandler.g
...
                         //remove possible old NBTTagCompound
                         this.mOutputItems[0].setTagCompound(new NBTTagCompound());
-                        GT_Utility.ItemNBT.setBookTitle(this.mOutputItems[0], s + " Construction Data");
+                        GT_Utility.ItemNBT.setBookTitle(this.mOutputItems[0], outputDisplayName + " Construction Data");
 
-                        NBTTagCompound tNBT = this.mOutputItems[0].getTagCompound();
-                        if (tNBT == null) {
-                            tNBT = new NBTTagCompound();
-                        }
+                        NBTTagCompound tNBT = ObjectUtils.defaultIfNull(this.mOutputItems[0].getTagCompound(), new NBTTagCompound());
```
Comment (Skiftcha) [inline src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java]: potentially unwanted garbage creation

Comment (Elisis) [reply src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java]: That's the old commit. Now Optional.ofNullable().orElse() is being used.

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java [hunk 3]
```diff
@@ -51,122 +53,162 @@ public MetaTileEntity newMetaEntity(IGregTechTileEntity aTileEntity) {
...
                 }
             }
             if (ItemList.IC2_Crop_Seeds.isStackEqual(aStack, true, true)) {
-                NBTTagCompound tNBT = aStack.getTagCompound();
-                if (tNBT == null) {
-                    tNBT = new NBTTagCompound();
-                }
+                
+            	NBTTagCompound tNBT = Optional.ofNullable(
+            	        aStack.getTagCompound()).orElse(new NBTTagCompound());
```
Comment (Skiftcha) [inline src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java:108]: potentially unwanted garbage creation

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java [hunk 4]
```diff
@@ -191,124 +238,138 @@ else if (aStack.getItemDamage() == 2 && aStack.toString().equals(GT_ModHandler.g
...
-                        String s = tRecipe.mOutput.getDisplayName();
-                        if (FMLCommonHandler.instance().getEffectiveSide().isServer()) {
-                            s = GT_Assemblyline_Server.lServerNames.get(tRecipe.mOutput.getDisplayName());
-                            if (s == null)
-                                s = tRecipe.mOutput.getDisplayName();
-                        }
+                        String outputDisplayName = tRecipe.mOutput.getDisplayName();
+                        if (FMLCommonHandler.instance().getEffectiveSide().isServer()) 
+                            outputDisplayName = Optional.ofNullable(
+                                    GT_Assemblyline_Server.lServerNames.get(tRecipe.mOutput.getDisplayName())).orElse(tRecipe.mOutput.getDisplayName());
```
Comment (Skiftcha) [inline src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java:292]: getOrDefault(outputDisplayName, outputDisplayName)

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java (hunk 2 above)
Comment (Skiftcha) [reply src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java]: yeah i saw so made new review

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java (hunk 3 above)
Comment (Skiftcha) [reply src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java:108]: so maybe Optional.ofNullable(this.mOutputItems[0].getTagCompound()).orElseGet(NBTTagCompound::new)

Comment (Elisis) [reply src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java:108]: The garbage creation was there before the changes, but yeah, using a Supplier does sound like a better idea

Comment (Skiftcha) [reply src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java:108]: it was not. if tNBT is not null it was not creating new instance of NBTTagCompound

Comment (Elisis) [reply src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java]: Ah, you're right. Regardless, it's fixed now.

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_CuringOven.java [hunk 5]
```diff
@@ -55,30 +55,48 @@ public Object getClientGUI(int aID, InventoryPlayer aPlayerInventory, IGregTechT
...
+                    	long tWorldTime = getBaseMetaTileEntity().getWorld().getTotalWorldTime();
+                        
+                    	aNBT = aNBT.getCompoundTag("GT.ToolStats");
+                        if (aNBT != null && aNBT.hasKey("Heat")) {
+                            
+                            tHeat = aNBT.getInteger("Heat");
+                            if(aNBT.hasKey("HeatTime")){
+                            	
+                                long tHeatTime = aNBT.getLong("HeatTime");
+                            	if (tWorldTime > (tHeatTime + 10)){
```
Comment (Skiftcha) [inline src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_CuringOven.java:78]: something strange with formatting here

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_CuringOven.java [hunk 6]
```diff
@@ -55,30 +55,48 @@ public Object getClientGUI(int aID, InventoryPlayer aPlayerInventory, IGregTechT
...
                         tHeat++;
-                        if(aNBT!=null){
-                        	aNBT.setInteger("Heat", tHeat);
-                        	aNBT.setLong("HeatTime", tWorldTime);}
-                        if(tHeat>GT_MetaGenerated_Tool.getPrimaryMaterial(tStack).mMeltingPoint){mInventory[0]=null;}
+                        if (aNBT != null) {
+                            
+                            aNBT.setInteger("Heat", tHeat);
+                            aNBT.setLong("HeatTime", tWorldTime);
+
```
Comment (Skiftcha) [inline src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_CuringOven.java:94]: adding blank strings here (and many other places) does not increase readability

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java [hunk 7]
```diff
@@ -51,122 +53,162 @@ public MetaTileEntity newMetaEntity(IGregTechTileEntity aTileEntity) {
...
-                        if (((IIndividual) tIndividual).analyze()) {
-                            getFillableStack().amount -= 100;
+        
+    	} else if ((GT_Utility.isStackValid(aStack)) && (aStack.stackSize > 0)) {
+            
+    		if ((this.getFillableStack() != null) && (this.getFillableStack().containsFluid(Materials.Honey.getFluid(100L)))) {
+                
+    			try {
+                	
+                	Object tIndividual = AlleleManager.alleleRegistry.getIndividual(aStack);
```
Comment (Skiftcha) [inline src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java:68]: and in this file

**Comment by Dream-Master** [comment]
@Elisis why you close this and not revolved issue and work on it ? Or is there no use ?

**Pull request closed without merging**